Sources/PackageLoading/ToolsVersionLoader.swift

```diff
@@ -88,32 +88,122 @@ extension Manifest {
 }
 
 public class ToolsVersionLoader: ToolsVersionLoaderProtocol {
-
+    
+    // FIXME: Remove this property and the initializer?
+    // Arbitrary tools versions are used only in `ToolsVersionLoaderTests.testVersionSpecificManifestFallbacks()`.
     let currentToolsVersion: ToolsVersion
 
     public init(currentToolsVersion: ToolsVersion = .currentToolsVersion) {
         self.currentToolsVersion = currentToolsVersion
     }
 
+    // FIXME: Use generic associated type `T: StringProtocol` instead of concrete types `String` and `Substring`, when/if this feature comes to Swift.
```

Comment:
I think the convention in this case is usually to make it a String, though that does sometimes require explicit creation of a String.

Reply:
All the new error cases I added use `Substring` as associated type, in order to work more seamlessly with manifest-splitting, and to avoid additional allocations. Should I change them to using `String` instead?

Reply:
I think the standard way to do the enums is usually to have the payload be `String`, and then using `String(...)` at creation time to wrap anything that isn't a `String` in a `String`.  This lets any clients use them as strings without impedance mismatch.  So I think that changing them to `String` might be the right thing here.  But I think you're right that it would be nice if it were possible to make them more generic, but it sounds as if we'd need stdlib changes for that.

Reply:
I think you're right about using `String` instead of `Substring`. `String` is a currency type in Swift while `Substring` isn't, so `String` is better. I changed the enums' associated value to `String`s in a50f997ae8b6a78a48e62cdeaf9f3c61dcf4de41 and 532a3157a75a3cdee2daaa19853c20490004cce7.

Reply:
Great, thanks for changing this.